### PR TITLE
Openstack provider

### PR DESCRIPTION
**Pull request by zreigz**
Our pull request delivers solution to create Kubernetes cluster on the top of OpenStack. Heat OpenStack Orchestration engine describes the infrastructure for Kubernetes cluster. CentoOS images are used for Kubernetes host machines.

We tested our solution with DevStack and Citycloud provider.

We believe that our solution will fill the gap that which is on the market.

---

This change is [Reviewable](http://reviewable.k8s.io/reviews/kubernetes/kubernetes/21737)

### CODE REVIEW

**Comment by zreigz** [comment]
@ohtake is also a author of those commits

**Comment by davidopp** [comment]
ok to test

**Comment by davidopp** [comment]
@ohtake needs to sign the CLA.

**Comment by davidopp** [comment]
cc/ @ashw7n

**Comment by ohtake** [comment]
@davidopp I have just singned the CLA. Could you re-verify the CLA?

**Comment by bgrant0607** [comment]
I don't think the CLA bot ever works for PRs with multiple authors.

**Comment by bgrant0607** [comment]
We're still trying to get 1.2 out the door and have a couple other new cloudproviders to review, so this will need to wait for a little while. 

It would also be useful to have some people using K8s on Openstack to review. I suggest posting to kubernetes-dev to plead for reviewers.

**Comment by zreigz** [comment]
I've created post on kubernetes-der regarding to openstack-provider:
https://groups.google.com/forum/#!topic/kubernetes-dev/t1-0zG3Lmeg

I hope someone will be interested.

**Comment by zreigz** [comment]
Ok thanks

**Comment by bgrant0607** [comment]
Would it be possible to split this into 2 PRs, each submitted by its author?

**Comment by idvoretskyi** [comment]
I suppose the target directory has to be renamed (or pushed into the new subdirectory of "openstack" directory).

The described method is not the only way to deploy Kubernetes cluster on OpenStack (don't forget about Murano, Magnum as well as about some minor projects), therefore it may confuse potential end-users.

**Comment by ohtake** [comment]
@bgrant0607 Okay, I will create one after we pick a directory name.

@idvoretskyi Good point. How about `openstack-heat` and `openstack/heat`? I prefer the former for simple hierarchy. What do you think?

**Comment by idvoretskyi** [comment]
@ohtake yes, your proposal if useful - let's do it.

**Comment by davidopp** [comment]
cc/ @mikedanese

**Comment by elsonrodriguez** [comment]
Tried running through this provisioner, ended up not getting past this portion of the script:
```
heat --api-timeout 60 stack-create -P external_network=public -P ssh_key_name=kubernetes_keypair -P server_image=30ecf289-c028-4a7c-b782-12fe3e20c95b -P master_flavor=m1.small -P minion_flavor=m1.small -P number_of_minions=1 -P max_number_of_minions=1 -P dns_nameserver=8.8.8.8 -P kubernetes_salt_url=http://.../kubernetes/kubernetes-salt.tar.gz -P kubernetes_server_url=http://.../kubernetes/kubernetes-server.tar.gz --template-file kubecluster.yaml KubernetesStack
b'ERROR: Property error: : resources.provision_network_master.properties.config: : Value must be a string'
```
This is with python-heatclient 1.1.0, 1.0.0, and Python 3.4.4 running from OSX.

**Comment by ohtake** [comment]
@elsonrodriguez
Thank you for trying the provider.

I could not reproduce it with CentOS 7 64bit, Python **2.7.5**, python-heatclient 1.1.0,
but I could reproduce it with CentOS 7 64bit, Python **3.4.3**, python-heatclient 1.1.0.

I think there is an issue in python-heatclient.
<https://bugs.launchpad.net/python-heatclient/+bug/1498037>
When I execute `heat -d stack-create ...` under Python 3, I got integer arrays.

**Comment by elsonrodriguez** [comment]
@ohtake Thanks, python version change fixed that issue. Will keep testing provider, but before I forget I had to change this line (wasn't sure of original intent, guessed an operator).
```
--- a/cluster/openstack/util.sh
+++ b/cluster/openstack/util.sh
@@ -91,7 +91,7 @@ function create-stack() {
   echo "[INFO] Execute commands to create Kubernetes cluster"
   # It is required for some cloud provider like CityCloud where swift client has different credentials
   source "${ROOT}/openrc-swift.sh"
-  if [[ ! -v OS_PROJECT_ID ]]; then
+  if [[ ! -z OS_PROJECT_ID ]]; then
     SWIFT_PROJECT_ID="${OS_TENANT_ID}"
   else
     SWIFT_PROJECT_ID="${OS_PROJECT_ID}"
```

Otherwise I would get a bash error:

```
./../cluster/../cluster/openstack/util.sh: line 94: conditional binary operator expected
```

This is with bash 3.2.57 on OSX.

**Comment by ohtake** [comment]
@elsonrodriguez
Reproduced with bash 4.1.2 on CentOS 6.7.
`-v` operator was [added in bash 4.2](http://stackoverflow.com/questions/3601515/how-to-check-if-a-variable-is-set-in-bash#comment31736775_17538964).

`! -z OS_PROJECT_ID` is always false because argument of `-z` is not varname but string.
I will modify the line `-z "${OS_PROJECT_ID}"`.

cc: @zreigz

**Comment by zreigz** [comment]
LGTM Thank for reporting this issue

**Comment by elsonrodriguez** [comment]
@zreigz @ohtake Got it working! 

It looks like the cloud provider isn't configured though. I'm going to look through the salt configuration and see if it's feasible. (a working configuration will look sort of like this: https://gist.github.com/elsonrodriguez/8d724baf910a5ad10a7e3b569b0f0b18)

I'm also behind a proxy, so I've had to make some hacks that I'd like to normalize and contribute.

How should we coordinate?

**Comment by elsonrodriguez** [comment]
@zreigz @ohtake I've gotten the initial support for proxies done: https://github.com/elsonrodriguez/kubernetes/commit/07e78fe2ca52a4d12ce47ea66105ba4684f2aa1d

I need to polish up a few things and add comments, and then I can move on to the cloudprovider.

I'll provide updates regarding the cloudprovider as I make progress.

**Comment by zreigz** [comment]
Thanks for your support

**Comment by elsonrodriguez** [comment]
@zreigz @ohtake The proxy branch is pretty much wrapped up, I need to do some more testing though. Also I'd like someone to test and see if it still works good on non-proxied setups: https://github.com/elsonrodriguez/kubernetes/commits/openstack-provider-proxy

I've also started on enabling cloudprovider support for Openstack. It passed a quick smoke-test(provisioning a cinder volume): https://github.com/elsonrodriguez/kubernetes/commit/0bf8b946022dfc0b46db5820bd87ee761ecf453e

Take a look and let me know what you think.

**Comment by elsonrodriguez** [comment]
@zreigz @ohtake 
I've made good progress on the cloud-provider integration with OpenStack: https://github.com/elsonrodriguez/kubernetes/commit/b4d36474292a86cf6fda7eb7ccda7d5579e0d05b

Also, here is an early preview of the documentation adjustments to reflect the changes I'm proposing, let me know if you like the direction: https://github.com/elsonrodriguez/kubernetes.github.io/commit/852b252361a02dcfc0aa39bd2a747d858c56948b

Give the changes a shot and let me know how it looks, or if you have any questions.

**Comment by elsonrodriguez** [comment]
@ohtake @zreigz I've sent a Pull Request for my proposed changes: https://github.com/FujitsuEnablingSoftwareTechnologyGmbH/kubernetes/pull/8

I think it's ready for testing and merging into this Pull Request so we can move forward and get more feedback.

I ended up having to modify the Salt code, so once my changes are merged in, a rebase will probably need to happen.

**Comment by elsonrodriguez** [comment]
@bgrant0607 I'm piling on contributions that affect commits from both @ohtake and @zreigz. How much worse have I made the CLA bot situation? How should I resolve it?

**Comment by elsonrodriguez** [comment]
Related Documentation PR here: https://github.com/FujitsuEnablingSoftwareTechnologyGmbH/kubernetes.github.io/pull/1/files?diff=split

**Comment by elsonrodriguez** [comment]
@ohtake @zreigz @cheld This PR was discussed during the Openstack Sig meeting: https://docs.google.com/document/d/1iAQ3LSF_Ky6uZdFtEZPD_8i6HXeFxIeW4XtGcUJtPyU/edit?ts=57150425#heading=h.1n8ybxoas1li (search for 'April 19th')

The main takeaways for this PR were that heat is a barrier to entry, and that a more monolithic CLI should be used.

So two main questions: 

- How feasible is it to do this without Heat?
- What's the most commonly used monolithic CLI we can use? (in please of nova, swift, glance)

**Comment by cheld** [comment]
@elsonrodriguez. Thanks for the information. Occasionally, I participate in this meeting. But, it is at midnight hours at local time :) and the agenda did not seem relevant to me...

Why is heat a problem? If we do it on our own we start to reimplement heat, don't we? And the provisioning process is run on client side. 

We could think about dropping the creation of the network and leave this for the end-user, and just create the master and the minions....

**Comment by xsgordon** [comment]
@cheld so at the core of the discussion was the goal of providing something in-tree in Kubernetes that is as basic a way as possible for someone from the Kubernetes community to get a K8S cluster up and running on an existing OpenStack cloud which they may or may not administer. If they don't and it doesn't have Heat then they are out of luck using this implementation. The feeling was that ideally the in-tree kube-up implementation for OpenStack should be as easy to use and widely deployable as possible.

For folks with more production-like requirements and/or broader flexibility as to which OpenStack projects to install the proposal was that we document the decision tree to help them navigate the other approaches (Heat, Murano, Magnum, etc.). The question posed on Slack was whether modifying these templates to use the basic client calls instead of doing them via Heat would be re-implementing Heat which is a fair question but it seems like only very basic orchestration is being performed here?. This may be more of an issue if the intent is to target a more production-like use case then the "just get me a cluster" user kicking the tires though there are Heat templates elsewhere [2][3][4] that more directly address this?

[1] https://www.openstack.org/assets/survey/April-2016-User-Survey-Report.pdf
[2] https://github.com/larsks/heat-kubernetes
[3] http://git.openstack.org/cgit/stackforge/heat-coe-templates/
[4] http://git.openstack.org/cgit/openstack/magnum/tree/magnum/templates/kubernetes

**Comment by cheld** [comment]
@xsgordon So the argument against heat is that not every cloud provider supports it yet, right?

I also think that this kube-up script based approach is rather for Kubernetes developers or people that compile Kubernetes themselves rather than production. 

Currently, we do not have the time to replace heat. What about restructuring the code so that heat could potentially be replaced?

**Comment by elsonrodriguez** [comment]
@cheld  If there is no time to replace heat, it's probably best to go forward as-is (no restructuring). Whoever performs the heat-removal work would be better off working from an unmolested base.

**Comment by elsonrodriguez** [comment]
@batikanu It may be the case that if you sign the CLA, the bot will be happy. Please sign the [Individual](http://code.google.com/legal/individual-cla-v1.0.html) or [Corporate](http://code.google.com/legal/corporate-cla-v1.0.html) CLA as soon as possible.

cc @cheld @ohtake @zreigz

**Comment by eparis** [comment]
to solve the CLA problem. Get everyone who wrote a commit to state, here in this PR, they are ok with the commit being merged into kubernetes. And admin will then add the label `cla: human-approved`. With that label you can ignore the `cla: no`

**Comment by batikanu** [comment]
I have signed the CLA already. Can you check again?

**Comment by zreigz** [comment]
After merge and rebase cluster is created but k8s doesn't work because of docker:
```
kwi 22 11:26:36 kubernetesstack-master kube-master-addons.sh[10621]: Failed to restart docker.service: Unit docker.service failed to load: No such file or directory.
```

**Comment by quinton-hoole** [comment]
Has anyone run the e2e conformance tests against this yet?

https://github.com/kubernetes/kubernetes/blob/c6e995a824094a96f7d43a25e897283f83a12997/docs/devel/e2e-tests.md . See the section on conformance tests.

**Comment by elsonrodriguez** [comment]
@zreigz This was due to a salt error. There was a reference to cadvisor that is no longer valid (it has been removed from the SaltStack).

This PR should fix it: https://github.com/FujitsuEnablingSoftwareTechnologyGmbH/kubernetes/pull/9

**Comment by zreigz** [comment]
Thanks for fix. I will test it on Monday. BTW now you can also run e2e tests.
```
 KUBERNETES_CONFORMANCE_TEST=y GINKGO_PARALLEL=y go run hack/e2e.go --v --test --test_args="--ginkgo.focus=\[Conformance\] --ginkgo.skip=\[Serial\]"
```

Before I merged your code it was 50/50. I will continue on Monday with your changes

**Comment by elsonrodriguez** [comment]
@zreigz @quinton-hoole If run with my pending patch( https://github.com/FujitsuEnablingSoftwareTechnologyGmbH/kubernetes/pull/9 ) I get 74 passed, 18 failed: https://gist.githubusercontent.com/elsonrodriguez/4d221b20c4a0d1e8fbdbf218e9ef390d/raw/a29f0756eabced52ed74e2edc852d443c69fb27b/Openstack%2520Comformance%2520Tests

Two of those tests are most likely related to my proxy setup (Internet and Services test)

For comparison, the AWS provider does 76 passed, 16 failed: https://gist.githubusercontent.com/elsonrodriguez/4d221b20c4a0d1e8fbdbf218e9ef390d/raw/a29f0756eabced52ed74e2edc852d443c69fb27b/AWS%2520Comformance%2520Tests

There's room for improvement, but I think we're looking good for this PR.

Also, @ohtake @batikanu, per https://github.com/kubernetes/kubernetes/pull/21737#issuecomment-213094725 , If we can get all the committers to comment in this thread and provide consent to merge, we can bypass the CLA bot. (P.S., I consent)

**Comment by ohtake** [comment]
@elsonrodriguez
batikanu and I have already singed the CLA.

* <https://github.com/kubernetes/kubernetes/pull/21737#issuecomment-191648030>
* <https://github.com/kubernetes/kubernetes/pull/21737#issuecomment-213351669>

**Comment by elsonrodriguez** [comment]
@ohtake Everyone has signed the CLA, but per @eparis, explicit consent has to be put in this PR to override the bot.

Something like "I consent to have my commits merged in."

**Comment by ohtake** [comment]
Thank you for explanation, @elsonrodriguez

@eparis, I consent to have my commits merged in.

**Comment by zreigz** [comment]
@ohtake @elsonrodriguez I had to rebase the branch. There were some conflicts. Please take a look if everything looks good.

**Comment by cheld** [comment]
In the SIG-OpenStack meet-up there was some discussion and confusion about this PR. So, let me clarify: In Kubernetes project we have some existing OpenStack providers, but they seem half-baked. E.g. dashboard and add-ons are not installed, monitoring not setup, etc...

In this PR we tried to be generic as possible and be as complient with Kubernetes design concept as possible. Some people seem not to like parts of the provining concepts like salt or swift.

I share to some extend these concerns, however if we want to replace them we must do it for all providers - not only OpenStack. Actually, we are already working on improving the provisioning process, but this is not ready yet - hopefully Kubernetes 1.3.

About heat: I assume that those users that have access to it also would like to use it. In addition, it could be leveraged for an node-autoscaling that is similar to GCE. We will also add Monasca monitoring integration in a subsequent PR (also similar to GCE). 

Our overall goal is to give Kubernetes developers and testers a platform that offers a similar functionality as for GCE.

**Comment by elsonrodriguez** [comment]
@zreig @quinton-hoole The tests are not passing to the same degree as Amazon anymore (also, the tests are no longer returning consistent results on AWS or OpenStack, these were best of 5):

Openstack: https://gist.githubusercontent.com/elsonrodriguez/71be812acb47a88c85887fbda56c6eb7/raw/6b05d10c44c94e8737cd298831899c10a5fa7586/OpenStack%2520Results
Amazon: https://gist.githubusercontent.com/elsonrodriguez/71be812acb47a88c85887fbda56c6eb7/raw/6b05d10c44c94e8737cd298831899c10a5fa7586/Amazon%2520Results

67 passed for Openstack
72 passed for Amazon

In addition, after some more testing/investigation, here are the current functional issues with this PR:

- OS_AUTH_URL with SSL support: This provider fails on OpenStack installs that use https for the OpenStack api. This was due to some sort of symlink issue with CentOS. There is a pending patch from @francios on Slack to fix this [http://paste.ubuntu.com/16064177/]()
- DNS_SERVER UX issues: On Openstack installs that don't have Neutron with internal DNS, the DNS_SERVER must still be set to a DNS server that can resolve the nodes, otherwise some functionality may not work (`kubectl exec`, `kubectl logs`). Registering nodes with a `--hostname-override <ip address>` doesn't work because of a mismatch between the OpenStack API and the overriden hostname.
- More robust generation of credentials for basic_auth and kubectl based on common.sh

If these issues and test results are not considered fatal to this patch, we should merge in this PR, and tackle remaining problems in separate issues/PRs. Otherwise we're going to run into more rebasing/testing overhead as time goes on, especially since this PR now touches the SaltStack setup and parts of kube-up.

Something else to consider is that contributions to fix these issues have to be in the form of a PR to Fujitsu's PR. Fujitsu has been very responsive, but even so collaborative efforts can be delayed due to timezone differences.

**Comment by cheld** [comment]
@zreigz @ohtake can you also take a look

**Comment by ctrlaltdel** [comment]
A patch for the OS_AUTH_URL with SSL support issue mentionned by @elsonrodriguez is available here: https://github.com/FujitsuEnablingSoftwareTechnologyGmbH/kubernetes/pull/10

**Comment by zreigz** [comment]
There is something wrong after rebase. Provisionig process is not finished correctly for eaxample Docker is not installed

**Comment by zreigz** [comment]
OK I know what is wrong. Missing ``` ` ``` in file cluster/saltbase/salt/kube-apiserver/init.sls

**Comment by elsonrodriguez** [comment]
@zreigz @batikanu @ctrlaltdel Could you all comment something like

"I consent to have my commits merged in to Kubernetes."

Also add a cc to `@quinton-hoole` and `@eparis`.

This is needed for manual override of the CLA bot.

**Comment by zreigz** [comment]
I consent to have my commits merged in to Kubernetes.
CC @quinton-hoole @eparis

**Comment by ctrlaltdel** [comment]
I consent to have my commits merged in to Kubernetes.
CC @quinton-hoole @eparis

**Comment by idvoretskyi** [comment]
@zreigz please, update the documentation https://github.com/kubernetes/kubernetes.github.io/pull/74 with the actual modifications for this provider.

Thank you.

**Comment by elsonrodriguez** [comment]
@idvoretskyi Could you clarify the documentation request? What needs to be updated?

**Comment by xsgordon** [comment]
@cheld I personally do not strongly object to the use of Heat at this point, but my concern is that the more we rely on projects that are not widely deployed the less useful the provider implementation is. Along similar lines as far as I can tell Monasca isn't represented in OpenStack Foundation user survey responses at all (versus e.g. Ceilometer).

I think we agree that the target is Kubernetes developers and testers, where we disagree is whether those users need or want to set up their own OpenStack cloud to do their work versus being able to run against an existing OpenStack cloud where available. Increased reliance on less widely deployed projects will ensure that they only have one option (deploy their own).

**Comment by cheld** [comment]
@xsgordon I would suggest to add _optional_ dependencies even to rare projects like Monasca, LBaaS, ... in case they are required to get fully integrated Kubernetes experience. I agree that this mostly interesting for demo and (integration-)development and might not have that much practical value as of now.  

Heat is not optional and your argument that this provider will not work with all cloud providers is valid. On the other hand, heat is not rare and already supported by the majority. If heat is not acceptable, than we should close this PR, because replacement is not a quick patch. (However, in this case Kubernetes will continue to have no usable provider for OpenStack...)

**Comment by batikanu** [comment]
I consent to have my commits merged in to Kubernetes.
CC @quinton-hoole @eparis

**Comment by xsgordon** [comment]
@cheld it's not really my argument, I am reflecting the discussion from the SIG (as requested in https://github.com/kubernetes/kubernetes/pull/21737#issuecomment-212288290 ) where the majority of folks present raised the concern that the majority of production OpenStack clouds do not in fact support Heat (only 39%).

My POV is somewhat different than the majority as I work for a vendor whose distribution has been installing Heat OOTB since ~ Havana but none the less I am attempting to accurately convey what was discussed in the context of the SIG (which I should add I do not believe to be the result of confusion about this PR but rather disagreement about what a default provider impl. for OpenStack should look like).

In the context of the SIG we are currently trying to pin down what "ideal state" would look like in terms of optional services, networking, etc., I am attempting to frame a strawman proposal to get things started which you might like to contribute to also as it would like help guide next steps beyond this initial PR?

**Comment by cheld** [comment]
So, if I understand you correctly, you suggest to write a design document, first?

**Comment by elsonrodriguez** [comment]
@xsgordon , @cheld 

The 39% Heat number comes from this Dataset: https://www.openstack.org/assets/survey/April-2016-User-Survey-Report.pdf

From that document:

- 66% of production Openstack installs have Nova
- 57% have Neutron
- 62% have Glance
- 39% have Heat (as you mentioned)
- 32% have Swift

So Nova, which is the piece that provisions VMs, and what many people consider synonymous with private clouds, is **missing** from 1/3rd of OpenStack installs.

Neutron, which defines networks, and Glance, which handles the upload of OS images, are in a similar state.

Swift, which is analogous to Amazon's S3, and is typically used by kube-up to store k8s artifacts, is down even below Heat.

My point is that in order provide an experience similar to the GCE or AWS kube-up providers, some minimum requirements have to be accepted.

Now outside this scope I think the bar can be set right at ground level, and there is value in that, but the amount of work required to re-implement the basics of a private cloud (or work around the absence of such basics) are out of scope for this PR, and can be addressed in subsequent PRs or other projects (the working doc you mentioned).

**Comment by xsgordon** [comment]
@cheld not quite, I think I'm happy to accept the current requirements to get something going but what I am suggesting is further discussion in the context of the SIG on what should be mandatory versus optional going forward as the functionality is expanded.

@elsonrodriguez the percentages are I believe that of the respondents to the question who selected both the project and that they were operating a production environment but yes what I am endeavouring to highlight is that the further down that list you go the less likely the project is to be in an existing environment that someone might want to deploy K8S on.

When we include testing environments the basic infrastructure services receive high percentages overall (90+%) but similarly to the production-only numbers there is a very quick drop off and we see that even Swift (one of the original two projects) is only barely in the majority of respondents clouds overall (53%).

Monasca you will find does not appear at all (unless my eyes deceive me as while I expected a low number I was surprised by this) which I believe means no respondents selected it. According to the report 290 OpenStack users responded to this section of the survey (the survey gives different questions depending on whether you identify as a deployer, app developer, etc. up front).

My concern is simply ensuring that we are making K8S on OpenStack as accessible as possible both for initial deployment but also when making optional features available (for example based on available stats on deployment characteristics I would expect it to make more sense to base scaling logic on Ceilometer than Monasca for example - I believe both the heat-kubernetes and OpenShift-on-OpenStack projects have existing example code for this).

**Comment by elsonrodriguez** [comment]
@xsgordon Definitely noted. Past these current requirements we should definitely avoid adding features that require the less adopted OpenStack components. If such features are added, the logic should be cased out for when the components are not available.

@quinton-hoole @a-robinson @eparis We have written consent from all contributors so far, could one of you override the googlebot and set the `cla` label set to `yes`?

Also, if anyone else wants to review this provider and test the functionality that would be most welcome. See here for instructions:

https://github.com/FujitsuEnablingSoftwareTechnologyGmbH/kubernetes.github.io/blob/openstack-provider-doc/docs/getting-started-guides/openstack-heat.md

You'll have to checkout this branch of kubernetes to test: 
```
git fetch origin pull/21737/head:openstack-provider
git checkout openstack-provider
```

File: cluster/openstack-heat/util.sh [hunk 1]
```diff
@@ -32,7 +32,7 @@ fi
 # Verify prereqs on host machine
 function verify-prereqs() {
  # Check the OpenStack command-line clients
- for client in swift glance nova heat;
+ for client in swift glance nova openstack;
```
Comment (xsgordon) [inline cluster/openstack-heat/util.sh]: Discussed with @elsonrodriguez on Slack but I believe you still need Heat here or it results in:

openstack: 'stack' is not an openstack command. See 'openstack --help'.
Did you mean one of these?
  backup create
  backup delete
  backup list
  backup restore
  backup show

File: cluster/openstack-heat/util.sh [hunk 2]
```diff
@@ -0,0 +1,274 @@
...
+source "${ROOT}/${KUBE_CONFIG_FILE:-"config-default.sh"}"
+source "${KUBE_ROOT}/cluster/common.sh"
+if [ $CREATE_IMAGE = true ]; then
+source "${ROOT}/config-image.sh"
+fi
+
+# Verify prereqs on host machine
+function verify-prereqs() {
+ # Check the OpenStack command-line clients
+ for client in swift glance nova openstack;
```
Comment (xsgordon) [inline cluster/openstack-heat/util.sh]: Discussed with @elsonrodriguez on Slack but I believe you still need Heat here or it results in:

openstack: 'stack' is not an openstack command. See 'openstack --help'.
Did you mean one of these?
backup create
backup delete
backup list
backup restore
backup show

**Comment by quinton-hoole** [comment]
LGTM!  Thanks for your patience.  I assume verify-prereqs improvement suggested by @xsgordon can be added in a separate PR.

File: cluster/openstack-heat/util.sh (hunk 2 above)
Comment (idvoretskyi) [inline cluster/openstack-heat/util.sh]: @elsonrodriguez why do we need this specific **python-openstackclient** version?

Are there any restrictions with more recent versions of the client?

Comment (elsonrodriguez) [inline cluster/openstack-heat/util.sh]: @idvoretskyi There's some sort of subcommand difference depending on what version of openstackclient you're using. @xsgordon is the one who ran into the bug, they'll have more details.

No matter what, we are going to have to prescribe a specific version of openstackclient to make sure this works. The individual commands(heat, swift, nova, etc) appear to have more stable arguments and flags. However there were concerns about them being deprecated, and indeed in the most recent versions of the individual commands a huge, distracting deprecation notice was displayed every time a command was issued.

There will always be restrictions for the version of the client, and the restrictions are listed in the documentation, and SHOULD in the future be checked by this provider.

Comment (xsgordon) [inline cluster/openstack-heat/util.sh]: @elsonrodriguez yeah, I plan to create a PR for a check shortly after this lands.

**Comment by zreigz** [comment]
This PR and documentation https://github.com/kubernetes/kubernetes.github.io/pull/74 ready for review
PTAL

**Comment by elsonrodriguez** [comment]
@quinton-hoole Looks like the bot didn't like you :)

**Comment by quinton-hoole** [comment]
@elsonrodriguez That only because commits were added after I gave the LGTM. So bot is being good :-)  New commits also LGTM.  Re-adding label.

**Comment by quinton-hoole** [comment]
@elsonrodriguez Oh, I see what you mean, the bot reassigned it.  That was due to a bug in github a week or so ago.  I've manually re-assigned back to myself.

**Comment by idvoretskyi** [comment]
@zreigz @elsonrodriguez it is not mandatory, but nice to have:
1) requirements.txt file with the desired `python-'project'client` versions (to install it easily with `pip install -r requirements.txt` command/update the docs with this as well;
2) Replace "OpenStack" with "OpenStack-Heat" on  https://github.com/FujitsuEnablingSoftwareTechnologyGmbH/kubernetes/blob/openstack-provider/cluster/get-kube.sh#L41 as we are expecting other different OpenStack-related providers.

Everything else LGTM. @quinton-hoole, please, take a look, possibly you have some other/extra suggestions.

**Comment by zreigz** [comment]
Thanks for suggestions.  Regarding to:
1) It sounds as a good enhancement for the future. For sure we will do it.
2) Done 
3) **minion** is used as user for instances. It is not only matter of documentation. 
https://github.com/FujitsuEnablingSoftwareTechnologyGmbH/kubernetes/blob/openstack-provider/cluster/openstack-heat/kubernetes-heat/fragments/kube-user.yaml#L4
Finally we get Kubernetes cluster, so is good to keep k8s name convention.

**Comment by elsonrodriguez** [comment]
@quinton-hoole Another commit snuck in, could you take another look?

**Comment by idvoretskyi** [comment]
@zreigz 
3) please, change that. Don't let people be confused.

**Comment by elsonrodriguez** [comment]
@zreigz @idvoretskyi That is a minor cosmetic issue. Almost every other kube-up provider still makes a reference to minions in their usage/code, along with the salt manifests.

I do think the names could be normalized, but that is squarely in the territory of a future PR.

**Comment by idvoretskyi** [comment]
@elsonrodriguez @zreigz agree.
LGTM

**Comment by pwittrock** [comment]
@zreigz 
@quinton-hoole 

Would you provide an update on the status + ETA for the documentation for this feature as well as add any PRs as they are created?

Not Started / In Progress / In Review / Done

Expected Merge Time:

Thanks
@pwittrock

**Comment by elsonrodriguez** [comment]
@pwittrock 
@zreigz 
@quinton-hoole 

Documentation is here: http://kubernetes.io/docs/getting-started-guides/openstack-heat/

And I have a pending PR to fix issues here: https://github.com/kubernetes/kubernetes.github.io/pull/534

Any review/testing would be appreciated.

**Comment by pwittrock** [comment]
@elsonrodriguez Is this already available or will it be available in v1.3?  The docs you linked are live, so it may be confusing to users if the feature is not available in the current release.

**Comment by elsonrodriguez** [comment]
@pwittrock 1.3 only, I can put a warning regarding the `curl | bash` pattern, and say that only 1.3 releases will work for the tar pattern.

**Pull request merged**